Rosserial

**[ERROR] [1684614705.969329]: Unable to sync with device; possible link problem or link software version mismatch such as hydro rosserial_python with groovy Arduino**

--- a/Hardware/IMU/IMU_2023/README.md
+++ b/Hardware/IMU/IMU_2023/README.md
@@ -1,1 +1,17 @@
-<Esquemáticos e informações sobre os IMUS>
+# Esquemáticos e informações sobre os IMUS
+ As bibliotecas Podem ser encontradas na própria Arduino IDE
+- Para o MPU6050: FastIMU
+- Para o BNO080: SparkFun BNO080 Cortex Based IMU
+
+## Lembrando que o as pinos padrão da ESP para I2C são:
+### D21 - SDA
+### D22 - SCL
+
+
+![image](https://github.com/HumanoidPequi/HumanoidPMec_LARC2023/assets/60831651/ec226b15-0de1-47e2-bb14-a14f83d5f3a4)
+
+## Esquemáticos (BNO080 e MPU6050)
+![image](https://github.com/HumanoidPequi/HumanoidPMec_LARC2023/assets/60831651/3da8655d-b3f6-46ab-bf41-bc29d33af83b)
+
+# ...[Atenção]...
+caso seja encontrado problemas com a utilização do MPU6050, adicione resistores de 10k nos sinais SDA e SCL em configuração pull-up (resistores ligados diretamente no positivo [3.3V] de alimentação)

--- a/Hardware/IMU/IMU_2023/README.md
+++ b/Hardware/IMU/IMU_2023/README.md
@@ -1,0 +1,1 @@
+<Esquemáticos e informações sobre os IMUS>

--- a/Hardware/IMU/IMU_2023/ROSBNO080/ROSBNO080.ino
+++ b/Hardware/IMU/IMU_2023/ROSBNO080/ROSBNO080.ino
@@ -1,0 +1,81 @@
+// Código quase completamente baseado no do Luis Fernando Ferreira
+// Pequi - Humanoide - Rui Jr
+
+// ROSS STUFF
+#include <ros.h>
+#include <ros/time.h>
+#include <std_msgs/Header.h>
+#include <sensor_msgs/Imu.h>
+#include <geometry_msgs/Quaternion.h>
+#include <geometry_msgs/Vector3.h>
+
+// IMU STUFF
+#include <Wire.h>
+#include "SparkFun_BNO080_Arduino_Library.h"
+BNO080 myIMU;
+
+// LED
+#define blinkPin 2
+
+// ROSS CONFIGS
+ros::NodeHandle nh;
+sensor_msgs::Imu inertial;
+geometry_msgs::Vector3 lin;
+geometry_msgs::Vector3 ang;
+ros::Publisher pub("micro/IMU", &inertial);
+char in[] = "/imu";
+
+void setup()
+{
+    pinMode(blinkPin, OUTPUT);
+    Wire.begin();
+    digitalWrite(blinkPin, LOW);
+    if (myIMU.begin() == false)
+    {
+        digitalWrite(blinkPin, HIGH);
+        delay(1000);
+        digitalWrite(blinkPin, LOW);
+        delay(1000);
+        digitalWrite(blinkPin, HIGH);
+        while (1)
+            ;
+    }
+    delay(10);
+    Wire.setClock(400000); // Increase I2C data rate to 400kHz
+    delay(10);
+    myIMU.enableLinearAccelerometer(8); // m/s^2 no gravity
+    myIMU.enableRotationVector(4);      // quat
+    myIMU.enableGyro(12);               // rad/s
+
+    nh.initNode();
+    nh.advertise(pub);
+
+    digitalWrite(blinkPin, LOW);
+}
+
+void loop()
+{
+    nh.spinOnce();
+    inertial.header.frame_id = in;
+    inertial.header.stamp = nh.now();
+    if (myIMU.dataAvailable() == true)
+    {
+        digitalWrite(blinkPin, HIGH);
+        // internal copies of the IMU data
+        byte linAccuracy = 0;
+        byte gyroAccuracy = 0;
+        float quatRadianAccuracy = 0;
+        byte quatAccuracy = 0;
+
+        // get IMU data in one go for each sensor type
+        myIMU.getLinAccel(lin.x, lin.y, lin.z, linAccuracy);
+        myIMU.getGyro(ang.x, ang.y, ang.z, gyroAccuracy);
+        myIMU.getQuat(inertial.orientation.x, inertial.orientation.y, inertial.orientation.z, inertial.orientation.w, quatRadianAccuracy, quatAccuracy);
+        inertial.angular_velocity = ang;
+        inertial.linear_acceleration = lin;
+        pub.publish(&inertial);
+        digitalWrite(blinkPin, LOW);
+    }
+    pub.publish(&inertial);
+    delay(1);
+}

--- a/Hardware/IMU/IMU_2023/ROSMPU6050/Madgwick.cpp
+++ b/Hardware/IMU/IMU_2023/ROSMPU6050/Madgwick.cpp
@@ -1,0 +1,249 @@
+//=============================================================================================
+// Madgwick.c
+//=============================================================================================
+//
+// Implementation of Madgwick's IMU and AHRS algorithms.
+// See: http://www.x-io.co.uk/open-source-imu-and-ahrs-algorithms/
+//
+// From the x-io website "Open-source resources available on this website are
+// provided under the GNU General Public Licence unless an alternative licence
+// is provided in source."
+//
+// Date			Author          Notes
+// 29/09/2011	SOH Madgwick    Initial release
+// 02/10/2011	SOH Madgwick	Optimised for reduced CPU load
+// 19/02/2012	SOH Madgwick	Magnetometer measurement is normalised
+//
+//=============================================================================================
+
+//-------------------------------------------------------------------------------------------
+// Header files
+
+#include "Madgwick.h"
+#include <math.h>
+#include <stdint.h>
+#include <Arduino.h>
+
+//-------------------------------------------------------------------------------------------
+// Definitions
+
+#define sampleFreqDef   512.0f          // sample frequency in Hz
+
+
+//============================================================================================
+// Functions
+
+//-------------------------------------------------------------------------------------------
+// AHRS algorithm update
+
+Madgwick::Madgwick() {
+	q0 = 1.0f;
+	q1 = 0.0f;
+	q2 = 0.0f;
+	q3 = 0.0f;
+	now = micros();
+}
+
+void Madgwick::update(float gx, float gy, float gz, float ax, float ay, float az, float mx, float my, float mz) {
+	float recipNorm;
+	float s0, s1, s2, s3;
+	float qDot1, qDot2, qDot3, qDot4;
+	float hx, hy;
+	float _2q0mx, _2q0my, _2q0mz, _2q1mx, _2bx, _2bz, _4bx, _4bz, _2q0, _2q1, _2q2, _2q3, _2q0q2, _2q2q3, q0q0, q0q1, q0q2, q0q3, q1q1, q1q2, q1q3, q2q2, q2q3, q3q3;
+
+	// Use IMU algorithm if magnetometer measurement invalid (avoids NaN in magnetometer normalisation)
+	if((mx == 0.0f) && (my == 0.0f) && (mz == 0.0f)) {
+		updateIMU(gx, gy, gz, ax, ay, az);
+		return;
+	}
+	
+	now = micros();
+	// Set integration time by time elapsed since last filter update
+	delta_t = ((now - last_update) / (float)1000000.0f);
+	last_update = now;
+
+	// Convert gyroscope degrees/sec to radians/sec
+	gx *= (float)0.0174533f;
+	gy *= (float)0.0174533f;
+	gz *= (float)0.0174533f;
+
+	// Rate of change of quaternion from gyroscope
+	qDot1 = (float)0.5f * (-q1 * gx - q2 * gy - q3 * gz);
+	qDot2 = (float)0.5f * (q0 * gx + q2 * gz - q3 * gy);
+	qDot3 = (float)0.5f * (q0 * gy - q1 * gz + q3 * gx);
+	qDot4 = (float)0.5f * (q0 * gz + q1 * gy - q2 * gx);
+
+	// Compute feedback only if accelerometer measurement valid (avoids NaN in accelerometer normalisation)
+	if(!((ax == (float)0.0f) && (ay == (float)0.0f) && (az == (float)0.0f))) {
+
+		// Normalise accelerometer measurement
+		recipNorm = invSqrt(ax * ax + ay * ay + az * az);
+		ax *= recipNorm;
+		ay *= recipNorm;
+		az *= recipNorm;
+
+		// Normalise magnetometer measurement
+		recipNorm = invSqrt(mx * mx + my * my + mz * mz);
+		mx *= recipNorm;
+		my *= recipNorm;
+		mz *= recipNorm;
+
+		// Auxiliary variables to avoid repeated arithmetic
+		_2q0mx = (float)2.0f * q0 * mx;
+		_2q0my = (float)2.0f * q0 * my;
+		_2q0mz = (float)2.0f * q0 * mz;
+		_2q1mx = (float)2.0f * q1 * mx;
+		_2q0 = (float)2.0f * q0;
+		_2q1 = (float)2.0f * q1;
+		_2q2 = (float)2.0f * q2;
+		_2q3 = (float)2.0f * q3;
+		_2q0q2 = (float)2.0f * q0 * q2;
+		_2q2q3 = (float)2.0f * q2 * q3;
+		q0q0 = q0 * q0;
+		q0q1 = q0 * q1;
+		q0q2 = q0 * q2;
+		q0q3 = q0 * q3;
+		q1q1 = q1 * q1;
+		q1q2 = q1 * q2;
+		q1q3 = q1 * q3;
+		q2q2 = q2 * q2;
+		q2q3 = q2 * q3;
+		q3q3 = q3 * q3;
+
+		// Reference direction of Earth's magnetic field
+		hx = mx * q0q0 - _2q0my * q3 + _2q0mz * q2 + mx * q1q1 + _2q1 * my * q2 + _2q1 * mz * q3 - mx * q2q2 - mx * q3q3;
+		hy = _2q0mx * q3 + my * q0q0 - _2q0mz * q1 + _2q1mx * q2 - my * q1q1 + my * q2q2 + _2q2 * mz * q3 - my * q3q3;
+		_2bx = sqrtf(hx * hx + hy * hy);
+		_2bz = -_2q0mx * q2 + _2q0my * q1 + mz * q0q0 + _2q1mx * q3 - mz * q1q1 + _2q2 * my * q3 - mz * q2q2 + mz * q3q3;
+		_4bx = (float)2.0f * _2bx;
+		_4bz = (float)2.0f * _2bz;
+
+		// Gradient decent algorithm corrective step
+		s0 = -_2q2 * ((float)2.0f * q1q3 - _2q0q2 - ax) + _2q1 * ((float)2.0f * q0q1 + _2q2q3 - ay) - _2bz * q2 * (_2bx * ((float)0.5f - q2q2 - q3q3) + _2bz * (q1q3 - q0q2) - mx) + (-_2bx * q3 + _2bz * q1) * (_2bx * (q1q2 - q0q3) + _2bz * (q0q1 + q2q3) - my) + _2bx * q2 * (_2bx * (q0q2 + q1q3) + _2bz * ((float)0.5f - q1q1 - q2q2) - mz);
+		s1 = _2q3 * ((float)2.0f * q1q3 - _2q0q2 - ax) + _2q0 * ((float)2.0f * q0q1 + _2q2q3 - ay) - (float)4.0f * q1 * (1 - (float)2.0f * q1q1 - (float)2.0f * q2q2 - az) + _2bz * q3 * (_2bx * ((float)0.5f - q2q2 - q3q3) + _2bz * (q1q3 - q0q2) - mx) + (_2bx * q2 + _2bz * q0) * (_2bx * (q1q2 - q0q3) + _2bz * (q0q1 + q2q3) - my) + (_2bx * q3 - _4bz * q1) * (_2bx * (q0q2 + q1q3) + _2bz * ((float)0.5f - q1q1 - q2q2) - mz);
+		s2 = -_2q0 * ((float)2.0f * q1q3 - _2q0q2 - ax) + _2q3 * ((float)2.0f * q0q1 + _2q2q3 - ay) - (float)4.0f * q2 * (1 - (float)2.0f * q1q1 - (float)2.0f * q2q2 - az) + (-_4bx * q2 - _2bz * q0) * (_2bx * ((float)0.5f - q2q2 - q3q3) + _2bz * (q1q3 - q0q2) - mx) + (_2bx * q1 + _2bz * q3) * (_2bx * (q1q2 - q0q3) + _2bz * (q0q1 + q2q3) - my) + (_2bx * q0 - _4bz * q2) * (_2bx * (q0q2 + q1q3) + _2bz * ((float)0.5f - q1q1 - q2q2) - mz);
+		s3 = _2q1 * ((float)2.0f * q1q3 - _2q0q2 - ax) + _2q2 * ((float)2.0f * q0q1 + _2q2q3 - ay) + (-_4bx * q3 + _2bz * q1) * (_2bx * ((float)0.5f - q2q2 - q3q3) + _2bz * (q1q3 - q0q2) - mx) + (-_2bx * q0 + _2bz * q2) * (_2bx * (q1q2 - q0q3) + _2bz * (q0q1 + q2q3) - my) + _2bx * q1 * (_2bx * (q0q2 + q1q3) + _2bz * ((float)0.5f - q1q1 - q2q2) - mz);
+		recipNorm = invSqrt(s0 * s0 + s1 * s1 + s2 * s2 + s3 * s3); // normalise step magnitude
+		s0 *= recipNorm;
+		s1 *= recipNorm;
+		s2 *= recipNorm;
+		s3 *= recipNorm;
+
+		// Apply feedback step
+		qDot1 -= beta * s0;
+		qDot2 -= beta * s1;
+		qDot3 -= beta * s2;
+		qDot4 -= beta * s3;
+	}
+
+	// Integrate rate of change of quaternion to yield quaternion
+	q0 += qDot1 * delta_t;
+	q1 += qDot2 * delta_t;
+	q2 += qDot3 * delta_t;
+	q3 += qDot4 * delta_t;
+
+	// Normalise quaternion
+	recipNorm = invSqrt(q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3);
+	q0 *= recipNorm;
+	q1 *= recipNorm;
+	q2 *= recipNorm;
+	q3 *= recipNorm;
+}
+
+//-------------------------------------------------------------------------------------------
+// IMU algorithm update
+
+void Madgwick::updateIMU(float gx, float gy, float gz, float ax, float ay, float az) {
+	float recipNorm;
+	float s0, s1, s2, s3;
+	float qDot1, qDot2, qDot3, qDot4;
+	float _2q0, _2q1, _2q2, _2q3, _4q0, _4q1, _4q2 ,_8q1, _8q2, q0q0, q1q1, q2q2, q3q3;
+
+	// Convert gyroscope degrees/sec to radians/sec
+	gx *= (float)0.0174533f;
+	gy *= (float)0.0174533f;
+	gz *= (float)0.0174533f;
+	
+	now = micros();
+	// Set integration time by time elapsed since last filter update
+	delta_t = ((now - last_update) / (float)1000000.0f);
+	last_update = now;
+
+	// Rate of change of quaternion from gyroscope
+	qDot1 = (float)0.5f * (-q1 * gx - q2 * gy - q3 * gz);
+	qDot2 = (float)0.5f * (q0 * gx + q2 * gz - q3 * gy);
+	qDot3 = (float)0.5f * (q0 * gy - q1 * gz + q3 * gx);
+	qDot4 = (float)0.5f * (q0 * gz + q1 * gy - q2 * gx);
+
+	// Compute feedback only if accelerometer measurement valid (avoids NaN in accelerometer normalisation)
+	if(!((ax == 0.0f) && (ay == 0.0f) && (az == 0.0f))) {
+
+		// Normalise accelerometer measurement
+		recipNorm = invSqrt(ax * ax + ay * ay + az * az);
+		ax *= recipNorm;
+		ay *= recipNorm;
+		az *= recipNorm;
+
+		// Auxiliary variables to avoid repeated arithmetic
+		_2q0 = (float)2.0f * q0;
+		_2q1 = (float)2.0f * q1;
+		_2q2 = (float)2.0f * q2;
+		_2q3 = (float)2.0f * q3;
+		_4q0 = (float)4.0f * q0;
+		_4q1 = (float)4.0f * q1;
+		_4q2 = (float)4.0f * q2;
+		_8q1 = (float)8.0f * q1;
+		_8q2 = (float)8.0f * q2;
+		q0q0 = q0 * q0;
+		q1q1 = q1 * q1;
+		q2q2 = q2 * q2;
+		q3q3 = q3 * q3;
+
+		// Gradient decent algorithm corrective step
+		s0 = _4q0 * q2q2 + _2q2 * ax + _4q0 * q1q1 - _2q1 * ay;
+		s1 = _4q1 * q3q3 - _2q3 * ax + 4.0f * q0q0 * q1 - _2q0 * ay - _4q1 + _8q1 * q1q1 + _8q1 * q2q2 + _4q1 * az;
+		s2 = (float)4.0f * q0q0 * q2 + _2q0 * ax + _4q2 * q3q3 - _2q3 * ay - _4q2 + _8q2 * q1q1 + _8q2 * q2q2 + _4q2 * az;
+		s3 = (float)4.0f * q1q1 * q3 - _2q1 * ax + 4.0f * q2q2 * q3 - _2q2 * ay;
+		recipNorm = invSqrt(s0 * s0 + s1 * s1 + s2 * s2 + s3 * s3); // normalise step magnitude
+		s0 *= recipNorm;
+		s1 *= recipNorm;
+		s2 *= recipNorm;
+		s3 *= recipNorm;
+
+		// Apply feedback step
+		qDot1 -= beta * s0;
+		qDot2 -= beta * s1;
+		qDot3 -= beta * s2;
+		qDot4 -= beta * s3;
+	}
+
+	// Integrate rate of change of quaternion to yield quaternion
+	q0 += qDot1 * delta_t;
+	q1 += qDot2 * delta_t;
+	q2 += qDot3 * delta_t;
+	q3 += qDot4 * delta_t;
+
+	// Normalise quaternion
+	recipNorm = invSqrt(q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3);
+	q0 *= recipNorm;
+	q1 *= recipNorm;
+	q2 *= recipNorm;
+	q3 *= recipNorm;
+}
+
+//-------------------------------------------------------------------------------------------
+// Fast inverse square-root
+// See: http://en.wikipedia.org/wiki/Fast_inverse_square_root
+
+float Madgwick::invSqrt(float x) {
+	float halfx = 0.5f * x;
+	float y = x;
+	long i = *(long*)&y;
+	i = 0x5f3759df - (i>>1);
+	y = *(float*)&i;
+	y = y * (1.5f - (halfx * y * y));
+	y = y * (1.5f - (halfx * y * y));
+	return y;
+}
+
+//-------------------------------------------------------------------------------------------

--- a/Hardware/IMU/IMU_2023/ROSMPU6050/Madgwick.h
+++ b/Hardware/IMU/IMU_2023/ROSMPU6050/Madgwick.h
@@ -1,0 +1,63 @@
+//=============================================================================================
+// Madgwick.h
+//=============================================================================================
+//
+// Implementation of Madgwick's IMU and AHRS algorithms.
+// See: http://www.x-io.co.uk/open-source-imu-and-ahrs-algorithms/
+//
+// From the x-io website "Open-source resources available on this website are
+// provided under the GNU General Public Licence unless an alternative licence
+// is provided in source."
+//
+// Date			Author          Notes
+// 29/09/2011	SOH Madgwick    Initial release
+// 02/10/2011	SOH Madgwick	Optimised for reduced CPU load
+//
+//=============================================================================================
+#ifndef __Madgwick_h__
+#define __Madgwick_h__
+#include <math.h>
+#include <stdint.h>
+
+//--------------------------------------------------------------------------------------------
+// Variable declaration
+class Madgwick {
+  private:
+    float delta_t = 0; // Used to control display output rate
+    uint32_t now = 0;        // used to calculate integration interval
+    uint32_t last_update = 0; // used to calculate integration interval
+    static float invSqrt(float x);
+    float beta;				// algorithm gain
+    float q0;
+    float q1;
+    float q2;
+    float q3;	// quaternion of sensor frame relative to auxiliary frame
+    
+
+    //-------------------------------------------------------------------------------------------
+    // Function declarations
+  public:
+    Madgwick(void);
+    void changeBeta(float newBeta) { beta = newBeta; }
+    void begin(float confBeta) { beta = confBeta; }
+    void update(float gx, float gy, float gz, float ax, float ay, float az, float mx, float my, float mz);
+    void updateIMU(float gx, float gy, float gz, float ax, float ay, float az);
+
+    float getQuatW() {
+      return q0;
+    }
+
+    float getQuatX() {
+      return q1;
+    }
+
+    float getQuatY() {
+      return q2;
+    }
+
+    float getQuatZ() {
+      return q3;
+    }
+
+};
+#endif

--- a/Hardware/IMU/IMU_2023/ROSMPU6050/ROSMPU6050.ino
+++ b/Hardware/IMU/IMU_2023/ROSMPU6050/ROSMPU6050.ino
@@ -1,0 +1,191 @@
+// PequiMec. - Humanoide - Rui Jr
+// ROSS STUFF
+#include <ros.h>
+#include <ros/time.h>
+#include <std_msgs/Header.h>
+#include <sensor_msgs/Imu.h>
+#include <geometry_msgs/Quaternion.h>
+#include <geometry_msgs/Vector3.h>
+
+// IMU STUFF
+#include <Wire.h>
+#include "FastIMU.h"
+#include "Madgwick.h"
+
+#define IMU_ADDRESS 0x68    // Change to the address of the IMU
+#define PERFORM_CALIBRATION // Comment to disable startup calibration
+MPU6050 IMU;                // Change to the name of any supported IMU!
+
+calData calib = {0}; // Calibration data
+AccelData IMUAccel;  // Sensor data
+GyroData IMUGyro;
+MagData IMUMag;
+Madgwick filter;
+
+ros::NodeHandle nh;
+
+sensor_msgs::Imu inertial;
+geometry_msgs::Vector3 lin;
+geometry_msgs::Vector3 ang;
+ros::Publisher pub("micro/IMU", &inertial);
+
+char in[] = "/imu";
+
+void setup()
+{
+
+    //
+    delay(1000);
+    Wire.begin();
+    Wire.setClock(400000); // 400khz clock
+
+    Serial.begin(115200);
+    while (!Serial)
+    {
+        ;
+    }
+
+    int err = IMU.init(calib, IMU_ADDRESS);
+    if (err != 0)
+    {
+        Serial.print("Error initializing IMU: ");
+        Serial.println(err);
+        while (true)
+        {
+            ;
+        }
+    }
+
+    if (err != 0)
+    {
+        Serial.print("Error Setting range: ");
+        Serial.println(err);
+        while (true)
+        {
+            ;
+        }
+    }
+
+#ifdef PERFORM_CALIBRATION
+    Serial.println("FastIMU Calibrated Quaternion example");
+    if (IMU.hasMagnetometer())
+    {
+        delay(1000);
+        Serial.println("Move IMU in figure 8 pattern until done.");
+        delay(3000);
+        IMU.calibrateMag(&calib);
+        Serial.println("Magnetic calibration done!");
+    }
+    else
+    {
+        delay(1000);
+    }
+    Serial.println("Keep IMU level.");
+    delay(5000);
+    IMU.calibrateAccelGyro(&calib);
+    Serial.println("Calibration done!");
+    Serial.println("Accel biases X/Y/Z: ");
+    Serial.print(calib.accelBias[0]);
+    Serial.print(", ");
+    Serial.print(calib.accelBias[1]);
+    Serial.print(", ");
+    Serial.println(calib.accelBias[2]);
+    Serial.println("Gyro biases X/Y/Z: ");
+    Serial.print(calib.gyroBias[0]);
+    Serial.print(", ");
+    Serial.print(calib.gyroBias[1]);
+    Serial.print(", ");
+    Serial.println(calib.gyroBias[2]);
+    if (IMU.hasMagnetometer())
+    {
+        Serial.println("Mag biases X/Y/Z: ");
+        Serial.print(calib.magBias[0]);
+        Serial.print(", ");
+        Serial.print(calib.magBias[1]);
+        Serial.print(", ");
+        Serial.println(calib.magBias[2]);
+        Serial.println("Mag Scale X/Y/Z: ");
+        Serial.print(calib.magScale[0]);
+        Serial.print(", ");
+        Serial.print(calib.magScale[1]);
+        Serial.print(", ");
+        Serial.println(calib.magScale[2]);
+    }
+    delay(5000);
+    IMU.init(calib, IMU_ADDRESS);
+
+    filter.begin(0.2f);
+#endif
+
+    //
+    nh.initNode();
+    nh.advertise(pub);
+}
+
+void loop()
+{
+    nh.spinOnce();
+    inertial.header.frame_id = in;
+    inertial.header.stamp = nh.now();
+
+    //
+    IMU.update();
+    IMU.getAccel(&IMUAccel);
+    IMU.getGyro(&IMUGyro);
+    if (IMU.hasMagnetometer())
+    {
+        IMU.getMag(&IMUMag);
+        filter.update(IMUGyro.gyroX, IMUGyro.gyroY, IMUGyro.gyroZ, IMUAccel.accelX, IMUAccel.accelY, IMUAccel.accelZ, IMUMag.magX, IMUMag.magY, IMUMag.magZ);
+    }
+    else
+    {
+        filter.updateIMU(IMUGyro.gyroX, IMUGyro.gyroY, IMUGyro.gyroZ, IMUAccel.accelX, IMUAccel.accelY, IMUAccel.accelZ);
+    }
+    Serial.print("QW: ");
+    Serial.print(filter.getQuatW());
+    Serial.print("\tQX: ");
+    Serial.print(filter.getQuatX());
+    Serial.print("\tQY: ");
+    Serial.print(filter.getQuatY());
+    Serial.print("\tQZ: ");
+    Serial.print(filter.getQuatZ());
+
+    IMU.getAccel(&IMUAccel);
+    Serial.print("\taccX: ");
+    Serial.print(IMUAccel.accelX);
+    Serial.print("\taccY");
+    Serial.print(IMUAccel.accelY);
+    Serial.print("\taccZ");
+    Serial.print(IMUAccel.accelZ);
+
+    IMU.getGyro(&IMUGyro);
+    Serial.print("\tgyroX");
+    Serial.print(IMUGyro.gyroX);
+    Serial.print("\tgyroY");
+    Serial.print(IMUGyro.gyroY);
+    Serial.print("\tgyroZ");
+    Serial.println(IMUGyro.gyroZ);
+
+    ang.x = IMUAccel.accelX;
+    ang.y = IMUAccel.accelY;
+    ang.z = IMUAccel.accelZ;
+
+    lin.x = IMUGyro.gyroX;
+    lin.y = IMUGyro.gyroY;
+    lin.z = IMUGyro.gyroZ;
+
+    inertial.angular_velocity = ang;
+    inertial.linear_acceleration = lin;
+
+    inertial.orientation.x = filter.getQuatX();
+    inertial.orientation.y = filter.getQuatY();
+    inertial.orientation.z = filter.getQuatZ();
+    inertial.orientation.w = filter.getQuatW();
+
+    pub.publish(&inertial);
+    pub.publish(&inertial);
+    delay(1);
+
+    delay(50);
+    //
+}

--- a/Hardware/controle_baixo_nivel.ino
+++ b/Hardware/controle_baixo_nivel.ino
@@ -1,6 +1,6 @@
 #include <Dynamixel2Arduino.h>
 #include <ros.h>
-#include <std_msgs/UInt16MultiArray.h>
+#include <std_msgs/Int16MultiArray.h>
 #include <ArduinoHardware.h>
 #define DXL_SERIAL Serial2
 const int DXL_DIR_PIN = 13; // DYNAMIXEL Shield DIR PIN
@@ -9,22 +9,24 @@ const uint8_t DXL_ID_CNT = 2;
 const uint8_t DXL_ID_LIST[DXL_ID_CNT] = {1, 2};
 const float DXL_PROTOCOL_VERSION = 1.0;
 
+const uint8_t base_servo[DXL_ID_CNT] = {180, 180};
+
 Dynamixel2Arduino dxl(DXL_SERIAL, DXL_DIR_PIN);
 
 ros::NodeHandle nh;
 
-void servo_cb(const std_msgs::UInt16MultiArray& cmd_msg) {
+void servo_cb(const std_msgs::Int16MultiArray& cmd_msg) {
   if (cmd_msg.data_length != DXL_ID_CNT) {
     return;
   }
 
   for (int i = 0; i < DXL_ID_CNT; i++) {
-    dxl.setGoalPosition(DXL_ID_LIST[i], cmd_msg.data[i], UNIT_DEGREE);
+    dxl.setGoalPosition(DXL_ID_LIST[i], base_servo[i] + cmd_msg.data[i], UNIT_DEGREE);
     delay(0.1);
   }
 }
 
-ros::Subscriber<std_msgs::UInt16MultiArray> sub("servo", servo_cb);
+ros::Subscriber<std_msgs::Int16MultiArray> sub("servo", servo_cb);
 
 void setup() {
   Serial.begin(57600);
@@ -39,6 +41,11 @@ void setup() {
     dxl.torqueOff(DXL_ID_LIST[i]);
     dxl.setOperatingMode(DXL_ID_LIST[i], OP_POSITION);
     dxl.torqueOn(DXL_ID_LIST[i]);
+  }
+  
+  for (int i = 0; i < DXL_ID_CNT; i++) {
+    dxl.setGoalPosition(DXL_ID_LIST[i], base_servo[i], UNIT_DEGREE);
+    delay(0.1);
   }
 }
 

--- a/Hardware/controle_baixo_nivel.ino
+++ b/Hardware/controle_baixo_nivel.ino
@@ -1,142 +1,48 @@
 #include <Dynamixel2Arduino.h>
-#include <actuator.h>
 #include <ros.h>
-#include <std_msgs/UInt16.h>
-#include <ArduinoHardware.h>;
-#define USE_USBCON
-#define DXL_SERIAL Serial
-const int DXL_DIR_PIN = 2; // DYNAMIXEL Shield DIR PIN
+#include <std_msgs/UInt16MultiArray.h>
+#include <ArduinoHardware.h>
+#define DXL_SERIAL Serial2
+const int DXL_DIR_PIN = 13; // DYNAMIXEL Shield DIR PIN
 
-//const uint8_t DXL_ID = 1;
-const uint8_t BROADCAST_ID = 254;
-const float DXL_PROTOCOL_VERSION = 2.0;
 const uint8_t DXL_ID_CNT = 2;
 const uint8_t DXL_ID_LIST[DXL_ID_CNT] = {1, 2};
-//const uint16_t user_pkt_buf_cap = 128;
-//uint8_t user_pkt_buf[user_pkt_buf_cap];
-
-/*// Starting address of the Data to read; Present Position = 132
-const uint16_t SR_START_ADDR = 132;
-// Length of the Data to read; Length of Position data of X series is 4 byte
-const uint16_t SR_ADDR_LEN = 4;
-// Starting address of the Data to write; Goal Position = 116
-const uint16_t SW_START_ADDR = 116;
-// Length of the Data to write; Length of Position data of X series is 4 byte
-const uint16_t SW_ADDR_LEN = 4;
-typedef struct sr_data{
-  int32_t present_position;
-} __attribute__((packed)) sr_data_t;*/
-typedef struct sw_data{
-  uint8_t* goal_position;
-} __attribute__((packed)) sw_data_t;
-
-
-/*sr_data_t sr_data[DXL_ID_CNT];
-DYNAMIXEL::InfoSyncReadInst_t sr_infos;
-DYNAMIXEL::XELInfoSyncRead_t info_xels_sr[DXL_ID_CNT];*/
-
-sw_data sw_data[DXL_ID_CNT];
-DYNAMIXEL::InfoSyncWriteInst_t sw_infos;
-DYNAMIXEL::XELInfoSyncWrite_t info_xels_sw[DXL_ID_CNT];
+const float DXL_PROTOCOL_VERSION = 1.0;
 
 Dynamixel2Arduino dxl(DXL_SERIAL, DXL_DIR_PIN);
 
-ros::NodeHandle  nh;
+ros::NodeHandle nh;
 
-uint8_t** goal_position;
-uint8_t goal_position_index = 0;
-uint8_t i;
-
-/*void servo_cb( const std_msgs::UInt16& cmd_msg){
-  dxl.setGoalPosition(DXL_ID_LIST, cmd_msg.data);  
-}*/ 
-
-  void servo_cb( const std_msgs::UInt16& cmd_msg){
-  // Fill the members of structure to syncWrite using internal packet buffer
-  //sw_infos.packet.p_buf = nullptr;
-  //sw_infos.packet.is_completed = false;
-  //sw_infos.addr = SW_START_ADDR;
-  //sw_infos.addr_length = SW_ADDR_LEN;
-  //sw_infos.p_xels = info_xels_sw;
-  sw_infos.xel_count = 0;
-  uint16_t conv_data = cmd_msg.data;
-  //goal_position = new uint8_t*[DXL_ID_CNT];
-   for (i = 0; i < DXL_ID_CNT; i++) {
-    goal_position[i] = new uint8_t[sizeof(uint16_t)];
-    memcpy(goal_position[i], &cmd_msg.data, sizeof(uint16_t));
+void servo_cb(const std_msgs::UInt16MultiArray& cmd_msg) {
+  if (cmd_msg.data_length != DXL_ID_CNT) {
+    return;
   }
 
-  for(i = 0; i < DXL_ID_CNT; i++){
-    info_xels_sw[i].id = DXL_ID_LIST[i];
-    info_xels_sw[i].p_data = (uint8_t*)&conv_data;
-    sw_infos.xel_count++;
-  }
-  sw_infos.is_info_changed = true;
-
-  static uint32_t try_count = 0;
-  uint8_t recv_cnt;
-  
-  // Insert a new Goal Position to the SyncWrite Packet
-  for(i = 0; i < DXL_ID_CNT; i++){
-    sw_data[i].goal_position = (uint8_t*)&goal_position[goal_position_index];
-  }
-
-  // Update the SyncWrite packet status
-  sw_infos.is_info_changed = true;
-
-  DXL_SERIAL.print("\n>>>>>> Sync Instruction Test : ");
-  DXL_SERIAL.println(try_count++);
-  
-  // Build a SyncWrite Packet and transmit to DYNAMIXEL  
-  if(dxl.syncWrite(&sw_infos) == true){
-    DXL_SERIAL.println("[SyncWrite] Success");
-    for(i = 0; i<sw_infos.xel_count; i++){
-      //DXL_SERIAL.print("  ID: ");DXL_SERIAL.println(sw_infos.p_xels[i].id);
-      //DXL_SERIAL.print("\t Goal Position: ");DXL_SERIAL.println(String(sw_data[i].goal_position));
-      DXL_SERIAL.print("Congratulations!");
-    }
-    if(goal_position_index == 0)
-      goal_position_index = 1;
-    else
-      goal_position_index = 0;
-  } else {
-    DXL_SERIAL.print("[SyncWrite] Fail, Lib error code: ");
-    DXL_SERIAL.print(dxl.getLastLibErrCode());
-  }
-  DXL_SERIAL.println();
-
-  delay(250);
   for (int i = 0; i < DXL_ID_CNT; i++) {
-    delete[] goal_position[i];
+    dxl.setGoalPosition(DXL_ID_LIST[i], cmd_msg.data[i], UNIT_DEGREE);
+    delay(0.1);
   }
-  delete[] goal_position;
 }
 
-ros::Subscriber<std_msgs::UInt16> sub("servo", servo_cb);
+ros::Subscriber<std_msgs::UInt16MultiArray> sub("servo", servo_cb);
 
 void setup() {
-  // put your setup code here, to run once:
-  dxl.begin(57600);
-  goal_position = new uint8_t*[DXL_ID_CNT];
-    for (int i = 0; i < DXL_ID_CNT; i++) {
-    goal_position[i] = new uint8_t[sizeof(uint16_t)];
-  }
+  Serial.begin(57600);
+  dxl.begin(1000000);
   nh.initNode();
   nh.subscribe(sub);
-  // Set Port Protocol Version. This has to match with DYNAMIXEL protocol version.
+
   dxl.setPortProtocolVersion(DXL_PROTOCOL_VERSION);
-  // Get DYNAMIXEL information
-  dxl.ping(DXL_ID_LIST[i]);
-  for(i = 0; i < DXL_ID_CNT; i++){ 
-    // Turn off torque when configuring items in EEPROM area
+
+  for (int i = 0; i < DXL_ID_CNT; i++) {
+    dxl.ping(DXL_ID_LIST[i]);
     dxl.torqueOff(DXL_ID_LIST[i]);
     dxl.setOperatingMode(DXL_ID_LIST[i], OP_POSITION);
+    dxl.torqueOn(DXL_ID_LIST[i]);
   }
-  dxl.torqueOn(BROADCAST_ID);
 }
-  
+
 void loop() {
-  // put your main code here, to run repeatedly:
   nh.spinOnce();
   delay(1);
 }

--- a/Hardware/low_level_control.ino
+++ b/Hardware/low_level_control.ino
@@ -1,0 +1,50 @@
+#define USE_USBCON
+#include <Dynamixel2Arduino.h>
+#include <actuator.h>
+#include <ros.h>
+#include <std_msgs/UInt16.h>
+#define DXL_SERIAL Serial2
+const int DXL_DIR_PIN = 13; // DYNAMIXEL Shield DIR PIN
+int LED_BUILTIN = 2; 
+const uint8_t DXL_ID = 1;
+const float DXL_PROTOCOL_VERSION = 2.0;
+
+Dynamixel2Arduino dxl(DXL_SERIAL, DXL_DIR_PIN);
+
+ros::NodeHandle  nh;
+
+void servo_cb( const std_msgs::UInt16& cmd_msg){
+  dxl.setGoalPosition(DXL_ID, cmd_msg.data);  
+}
+
+ros::Subscriber<std_msgs::UInt16> sub("servo", servo_cb);
+
+void setup() {
+  Serial.begin(115200);
+  while(!Serial);
+  Serial.write("hello");
+  // put your setup code here, to run once:
+  pinMode (LED_BUILTIN, OUTPUT); 
+  digitalWrite (LED_BUILTIN, HIGH);
+  nh.getHardware()->setBaud(115200);
+  dxl.begin(9600);
+  nh.initNode();
+  nh.subscribe(sub);
+  // Set Port Protocol Version. This has to match with DYNAMIXEL protocol version.
+  dxl.setPortProtocolVersion(DXL_PROTOCOL_VERSION);
+  // Get DYNAMIXEL information
+  dxl.ping(DXL_ID);
+
+  // Turn off torque when configuring items in EEPROM area
+  dxl.torqueOff(DXL_ID);
+  dxl.setOperatingMode(DXL_ID, OP_POSITION);
+  dxl.torqueOn(DXL_ID);  
+  delay(2000);
+}
+
+void loop() {
+  digitalWrite (LED_BUILTIN, LOW);
+  // put your main code here, to run repeatedly:
+  nh.spinOnce();
+  delay(1);
+}

